api/client/service/update.go

```diff
@@ -388,6 +388,17 @@ func portConfigToString(portConfig *swarm.PortConfig) string {
 	return fmt.Sprintf("%v/%s", portConfig.PublishedPort, protocol)
 }
 
+func validatePort(val string) (string, error) {
+	proto, port := nat.SplitProtoPort(val)
+	if proto != "tcp" && proto != "udp" {
+		return "", fmt.Errorf("invalid protocol '%s' for %s", proto, val)
+	}
+	if _, err := nat.ParsePort(port); err != nil {
+		return "", err
+	}
```

Comment:
I think we should have a special case here for `strings.Contains(port, ":")` and give a better error message.